Add LP SRE Cluster admin and RHMI SRE RBAC

alanmoran opened the pull request:
As per https://issues.redhat.com/browse/OSD-2783

- Create new layered-sre-cluster-admins (breakglass)
- Create CS SRE admins group (RHMI)
- New monitoring check for  layered-sre-cluster-admins group

alanmoran commented:
/hold

jewzaam commented on deploy/layered-sre-authorization/02-layered-cs-sre-admin.SubjectPermission.yaml:
```diff
@@ -0,0 +1,15 @@
+  - layered-cs-sre-admin
+  permissions:
+  - allowFirst: true
+    clusterRoleName: dedicated-admins-cluster
```
```suggestion
    clusterRoleName: layered-cs-sre-admin
```
Suspect this is a copy/paste error, intent is to authorize with the LP admin ClusterRole

jewzaam commented on deploy/layered-sre-authorization/02-layered-cs-sre-admin.SubjectPermission.yaml:
```diff
@@ -0,0 +1,15 @@
+  permissions:
+  - allowFirst: true
+    clusterRoleName: dedicated-admins-cluster
+    namespacesAllowedRegex: .*
```
```suggestion
    namespacesAllowedRegex: ^redhat-.*
```

Do you really need all customer namespaces?  You can limit this to the LP namespaces if not.

cblecker commented on deploy/layered-sre-authorization/00-layered-sre-cluster-admins.Group.yaml:
```diff
@@ -0,0 +1,6 @@
+apiVersion: user.openshift.io/v1
+kind: Group
+metadata:
+  name: layered-sre-cluster-admins
```
Should we prefix this with osd, like our other groups?

alanmoran replied:
Went the layered-* route so we don't confuse with SRE P groups. layered-* will be added to the webhook validation.

jewzaam commented:
Added support for `sss-config.yaml` in https://github.com/openshift/managed-cluster-config/pull/231
You can create this file in the directory with your resources and supply the RHMI label.
```
matchLabels:
    api.openshift.com/addon-rhmi-operator: "true"
```

alanmoran commented:
> Added support for `sss-config.yaml` in #231
> You can create this file in the directory with your resources and supply the RHMI label.
> 
> ```
> matchLabels:
>     api.openshift.com/addon-rhmi-operator: "true"
> ```

Thanks @jewzaam! I've applied the SSS config now. The build was also failing as I had added the RHMI_SRE_USER1 template to the selectorsyncset template directly. I've updated template.yaml to allow for these optional values.

jewzaam commented on deploy/layered-sre-authorization/02-layered-cs-sre-admin.SubjectPermission.yaml:
```diff
@@ -0,0 +1,15 @@
+  namespace: openshift-rbac-permissions
+spec:
+  clusterPermissions:
+  - dedicated-admins-cluster
```
```suggestion
  - layered-cs-sre-admin
```

This should not be the `dedicated-admins-cluster` ClusterRole

alanmoran replied:
I'm trying to achieve giving the group layered-cs-sre-admin the dedicated-admins-cluster role available clusterwide, and the layered-cs-sre-admin role available in all the non redhat namespaces. Am I applying this wrong?

jewzaam replied:
You could add layered-cs-sre-admin to this, it's an array.  Why does CS SRE need dedicated-admin permissions in cluster?  I expect your `layered-cs-sre-admin` ClusterRole to have everything CS SRE needs in a customer namespace and that those permissions are something less than dedicated-admin.

alanmoran replied:
Ok, Interesting. My view is that we have the same as the customer and a subset more. This is purely from the nature of the type of work we do. We are managing middleware products with customers more than on behalf. An example here is products dynamically creating namespaces that hold config. Customers can manage this but so can CS SRE.

jewzaam replied:
Ok, I see my disconnect.  You're ClusterRole reads to me like a project level permissions.  I missed the clusterPermissions will create the ClusterRoleBinding.  Your layered-cs-sre-admin ClusterRole is too broad to grant as a ClusterRoleBinding, please split it into two ClusterRoles, one for cluster scope resources (project, namespace, user, etc) and project resources (everything else).  In the SubjectPermissions the cluster scope ClusterRole goes into `clusterPermissions`, the project scope ClusterRole goes into `permissions`.

alanmoran replied:
I also see what you were originally asking.  I've splt the cluster scope and project resources out (layered-cs-sre-admin-cluster and layered-cs-sre-admin-project)

jewzaam commented on deploy/layered-sre-authorization/01-layered-cs-sre-admin.ClusterRole.yaml:
```diff
@@ -0,0 +1,87 @@
+- apiGroups:
+  - ""
+  resources:
+  - namespaces
```
Same issue as `projects` resource.

jewzaam commented on deploy/layered-sre-authorization/01-layered-cs-sre-admin.ClusterRole.yaml:
```diff
@@ -0,0 +1,87 @@
+- apiGroups:
+  - project.openshift.io
+  resources:
+  - projects
```
I believe this must be granted at a cluster scope.  If so, need to create a separate ClusterRole and a ClusterRoleBinding to the `layered-cs-sre-admins` group.

alanmoran replied:
I don't really understand what needs to be done here. This is a ClusterRole?

jewzaam replied:
A ClusterRole isn't applied until you bind it to something.  To grant project and namespace I think you have to use a ClusterRoleBinding.  Using SubjectPermission CR will create RoleBindings in the matching namespaces.  You need to create a ClusterRoleBinding as a part of this PR to get the cluster scope permissions.

For example, dedicated-admin permissions are split into two ClusterRoles: dedicated-admins-cluster [1] and dedicated-admins-project [2].  A ClusterRoleBinding is created for the dedicated-admins-cluster permissions [3], a SubjectPermission is created for the dedicated-admins-project permissions [4] which results in RoleBindings in each customer namespace.

[1] https://github.com/openshift/managed-cluster-config/blob/master/deploy/rbac-permissions-operator-config/00-dedicated-admins-cluster.ClusterRole.yaml
[2] https://github.com/openshift/managed-cluster-config/blob/master/deploy/rbac-permissions-operator-config/00-dedicated-admins-project.ClusterRole.yaml
[3] https://github.com/openshift/managed-cluster-config/blob/master/deploy/rbac-permissions-operator-config/00-dedicated-admins-cluster.ClusterRoleBinding.yaml
[4]

alanmoran replied:
I thought the RBAC operator created the ClusterRoleBinding too. I'll validate this tomorrow to check. I made a change to this piece earlier in the PR and now need to retest it.

alanmoran replied:
Ok, I've tested this. The ClusterRoleBinding is created (if missing). I did lose the layered-cs-sre-admin ClusterRoleBinding though. I've re-added it to the SubjectPermission. Nice catch. Thank you.

jewzaam commented on deploy/layered-sre-authorization/01-layered-cs-sre-admin.ClusterRole.yaml:
```diff
@@ -0,0 +1,87 @@
+  - namespaces/finalize
+  verbs:
+  - '*'
+# CS SRE can use oc debug
```
Using `oc debug` requires being run in the `default` namespace.  Need to split this out to a separate permissions.
See https://github.com/openshift/ops-sop/blob/b40edcb9c761d7023b7a8ced30cc9c7aea05cad1/v4/support.asciidoc#23-access-node-via-oc

alanmoran replied:
The intention for this is to run debug containers in RHMI namespaces and not against nodes. Not sure if I still need to split out. If I do, what permissions do I need to split this to? CS SRE have access to the default NS.

jewzaam replied:
Good point, I think "node" when I see `oc debug`.  Appears to work on a pod, so yes this is fine as-is.

jewzaam commented on deploy/layered-sre-authorization/01-layered-cs-sre-admin-cluster.ClusterRole.yaml:
```diff
@@ -0,0 +1,54 @@
+  verbs:
+  - get
+  - list
+  - watch
```
Why cluster-scope get/list/watch?  My concern with this is it enables access to all secrets in the cluster and I'd just like to check that this is actually needed.

jewzaam replied:
Discussed on slack.  Basically it's there to enable access to the aws creds in case it's needed.  Discussed with @alanmoran about this being a break-glass thing, CS SRE could elevate privileges and get access.  Making elevating required means some thought has to go into the actions that are highly privileged.

jewzaam replied:
```suggestion
```

jewzaam commented on hack/00-osd-managed-cluster-config.selectorsyncset.yaml.tmpl:
```diff
@@ -535,6 +551,159 @@ objects:
+        verbs:
+        - get
+        - list
+        - watch
```
```suggestion
```

jewzaam commented:
/approve
/lgtm

alanmoran commented:
/hold cancel

jewzaam commented:
/lgtm

jewzaam commented:
/retest
